### ISSUE TITLE
module api that can process options

### ISSUE THREAD

**Issue by Toxicable** [terser/terser#467]
# Feature request

## Problem

We're looking at making a terser wrapper that operates on all files in a directory.
Meaning it executes 1 terser per file in a directory, speeding up execution in situations with multiple files to minify at the same time.

Currently we're using a `child_process` approach, while this works reasonable well there's performance and memory overheads to doing this.
https://github.com/bazelbuild/rules_nodejs/blob/master/packages/terser/src/index.js

So instead we'd like to take the `worker_threads` approach instead.
https://github.com/bazelbuild/rules_nodejs/issues/1177
However, we've ran into a few issues with this
We want to reuse the code found https://github.com/terser/terser/blob/master/bin/terser#L65 to process the command line args.
In order to do that we simply call `require('terser/bin/terser)` which executes tersers CLI interface - this works fine in a child process, but not under a `worker_threads` since it's async and somtimes calls process.exit.

The solution here is for us to instead use the terser module API directly instead.
The issue with that is that it dosen't do all the processing on the options that's done by the binary file.

## Request

Move all options parsing and defaulting logic into module API, to reduce external overhead and maintain a consistant handling of args externally.

**Comment by kzc** [comment]
See: https://github.com/terser/terser/issues/75#issuecomment-530650363

All CLI options are available in the `minify` API. The API does not have any dependency on the filesystem by design so it can be used in browser or embedded scenarios.

Be aware that `minify` operations involving the name cache cannot be parallelized as `minify` must process every JS input sequentially in order to have consistency for all mangled names and properties across all inputs.

**Comment by Toxicable** [comment]
@kzc Indeed, I can see the api is the same, but there's extra logic going into parsing those CLI options that isn't there.
Such as this line here: https://github.com/terser/terser/blob/master/bin/terser#L66
or these here: https://github.com/terser/terser/blob/master/bin/terser#L114-L121

We'd effectively have to copy paste that ( https://github.com/terser/terser/blob/master/bin/terser ) entire file to implement the same API. Which means there's room for divergances in the CLI API itself which we don't want, and don't want to maintain.

**Comment by kzc** [comment]
The CLI is unimportant - it's just a consumer of the `minify` API and an example of how it can be used. The `minify` API doesn't change in backwards incompatible ways. Copy the few hundred or so lines of code from the CLI and change it to suit your build tool's configuration scheme. Once in your tool it will be the canonical way you use it there.

**Comment by fabiosantoscode** [comment]
I don't think it's a bad idea to do something like:

```
// warning: super hypothetical code
if (module.parent) {
  module.exports = { run, output }
} else {
  run(process.argv).then(output).catch(...)
}
```

It would make some tests a bit nicer, since currently we use the child_process module to spawn the CLI and test it, while we could really be calling an async function as many times as we like.

If as a side effect people can create their own wrapper modules I think that's not bad at all for the Terser ecosystem as a whole.
